File: pubnub-gson/src/main/java/com/pubnub/api/enums/PNPushType.java
package com.pubnub.api.enums;

public enum PNPushType {

    APNS("apns"),

    MPNS("mpns"),

    /**
     * Use FCM instead
     */
    @Deprecated
    GCM("gcm"),

    FCM("gcm"),

    APNS2("apns2");

    private final String value;

    PNPushType(String name) {
        value = name;
    }

    @Override
    public String toString() {
        return this.value;
    }
}

File: pubnub-kotlin/pubnub-kotlin-core-api/src/commonMain/kotlin/com/pubnub/api/enums/PNPushType.kt
package com.pubnub.api.enums

enum class PNPushType(private val value: String) {
    @Deprecated(
        replaceWith = ReplaceWith("FCM"),
        message = "GCM is deprecated. Use FCM instead."
    )
    GCM("gcm"),
    APNS("apns"),
    MPNS("mpns"),
    FCM("gcm"),
    APNS2("apns2"),
    ;

    fun toParamString(): String {
        return value.lowercase()
    }

    companion object {
        fun fromParamString(paramString: String): PNPushType {
            return entries.first { it.toParamString() == paramString }
        }
    }

    override fun toString(): String {
        return value
    }
}
